src/main/java/net/glowstone/block/blocktype/BlockHopper.java

```diff
@@ -50,10 +60,127 @@ public TileEntity createTileEntity(GlowChunk chunk, int cx, int cy, int cz) {
     public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face, ItemStack holding, Vector clickedLoc) {
         super.placeBlock(player, state, face, holding, clickedLoc);
         setFacingDirection(state, face.getOppositeFace());
+        state.getBlock().getWorld().requestPulse(state.getBlock());
     }
 
     @Override
     protected MaterialMatcher getNeededMiningTool(GlowBlock block) {
         return ToolType.PICKAXE;
     }
+
+    @Override
+    public void receivePulse(GlowBlock block) {
+        if (block.getTileEntity() == null) {
+            return;
+        }
+        TEHopper hopper = (TEHopper) block.getTileEntity();
+        pullItems(block, hopper);
+        if (!((Hopper) block.getState().getData()).isPowered()) {
+            pushItems(block, hopper);
+        }
+    }
+
+    @Override
+    public void onRedstoneUpdate(GlowBlock block) {
+        ((Hopper) block.getState().getData()).setActive(!block.isBlockPowered());
+    }
+
+    private void pullItems(GlowBlock block, TEHopper hopper) {
+        GlowBlock source = block.getRelative(BlockFace.UP);
+        if (source.getType() == null || source.getType() == Material.AIR) {
```

Comment:
Hoppers with slabs or signs above will not work. getType() will never return null, only air.